### PR TITLE
chore(web): add inline assets

### PR DESCRIPTION
**Pull request by Matrixbirds**
To fix image assets broken when the end network is unreachable.

### CODE REVIEW

**Comment by BlackHole1** [comment]
ci failed reason:

  1. `\x00`: https://github.com/vitejs/vite/pull/3436
  2. `regex`: forget `$`, e.g: `.*/\.svg$` or use `endWith`: `filePath.endWith(".svg");`

**Review by hyrious — DISMISSED** [review]

File: web/flat-web/scripts/vite-plugin-inline-assets.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    return encodeURIComponent(normalizedContent).replace(urlHexPairsRE, specialHexEncode);
+}
+
+export function inlineAssets(): Plugin {
+    return {
+        name: "inline:assets",
+        enforce: "pre",
+        async transform(_, id: string) {
+            const filePath = cleanUrl(id);
+            if (/.(svg|jpg|jpeg|png|gif)$/i.test(filePath)) {
```
Comment (BlackHole1) [inline web/flat-web/scripts/vite-plugin-inline-assets.ts]:
```suggestion
            if (/\.(svg|jpg|jpeg|png|gif)$/i.test(filePath)) {
```

File: web/flat-web/scripts/vite-plugin-inline-assets.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        }
+        case "%3A": {
+            return ":";
+        }
+        case "%2F": {
+            return "/";
+        }
+        default: {
+            return match.toLowerCase();
+        } // compresses better
```
Comment (BlackHole1) [inline web/flat-web/scripts/vite-plugin-inline-assets.ts]:
```suggestion
        }
```

File: web/flat-web/scripts/vite-plugin-inline-assets.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
+import type { Plugin } from "vite";
+import { promises as fsp } from "fs";
+import mime from "mime/lite";
+
+const doubleQuoteRE = /"/g;
+const whitespaceRE = /\s+/g;
+const urlHexPairsRE = /%[\dA-F]{2}/g;
```
Comment (BlackHole1) [inline web/flat-web/scripts/vite-plugin-inline-assets.ts]:
```suggestion
const urlHexPairsRE = /%\d[\dA-F]/g;
```

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

File: web/flat-web/scripts/vite-plugin-inline-assets.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    return {
+        name: "inline:assets",
+        enforce: "pre",
+        async transform(_, id: string) {
+            const filePath = cleanUrl(id);
+            if (/\.(svg|jpg|jpeg|png|gif)$/i.test(filePath)) {
+                const content = await fsp.readFile(filePath);
+                const url = filePath.endsWith(".svg")
+                    ? `data:image/svg+xml;utf8,${uriEncodeSvg(content.toString("utf-8"))}`
+                    : `data:${mime.getType(filePath)};base64,${content.toString("base64")}`;
```
Comment (BlackHole1) [inline web/flat-web/scripts/vite-plugin-inline-assets.ts]:
```suggestion
                const imageFileContent = await fsp.readFile(filePath);
                const dataURI = filePath.endsWith(".svg")
                    ? `data:image/svg+xml;utf8,${uriEncodeSvg(svgFileContent.toString("utf-8"))}`
                    : `data:${mime.getType(filePath)};base64,${svgFileContent.toString("base64")}`;
```

File: web/flat-web/scripts/vite-plugin-inline-assets.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            return match.toLowerCase();
+        } // compresses better
+    }
+}
+
+const cleanUrl = (url: string): string => url.replace(/#.*$/, "").replace(/\?.*$/, "");
+
+// Adopted from https://github.com/tigt/mini-svg-data-uri
+function uriEncodeSvg(content: string): string {
+    const normalizedContent = content.trim().replace(whitespaceRE, " ").replace(doubleQuoteRE, "'");
```
Comment (BlackHole1) [inline web/flat-web/scripts/vite-plugin-inline-assets.ts]:
```suggestion
    const normalizedContent = content.trim().replace(/\s+/g, " ").replace(/"/g, "'");
```

File: web/flat-web/scripts/vite-plugin-inline-assets.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+    return encodeURIComponent(normalizedContent).replace(urlHexPairsRE, specialHexEncode);
```
Comment (BlackHole1) [inline web/flat-web/scripts/vite-plugin-inline-assets.ts]:
```suggestion
    return encodeURIComponent(normalizedContent).replace(/%[\dA-F]{2}/g, specialHexEncode);
```

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

File: web/flat-web/scripts/vite-plugin-inline-assets.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        case "%2F": {
+            return "/";
+        }
+        default: {
+            return match.toLowerCase();
+        }
+    }
+}
+
+const cleanUrl = (url: string): string => url.replace(/#.*$/, "").replace(/\?.*$/, "");
```
Comment (BlackHole1) [inline web/flat-web/scripts/vite-plugin-inline-assets.ts]:
```suggestion
// e.g:
// flat/node_modules/electron/index.js?v=19cea64f => flat/node_modules/electron/index.js
const removeQueryString = (url: string): string => url.replace(/#.*$/, "").replace(/\?.*$/, "");
```

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

File: web/flat-web/scripts/vite-plugin-inline-assets.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+export function inlineAssets(): Plugin {
+    return {
+        name: "inline:assets",
+        enforce: "pre",
+        async transform(_, id: string) {
+            const filePath = removeQueryString(id);
+            if (/\.(svg|jpg|jpeg|png|gif)$/i.test(filePath)) {
+                const imageFileContent = await fsp.readFile(filePath);
+                const url = filePath.endsWith(".svg")
+                    ? svgToTinyDataUri(imageFileContent.toString("utf-8"), "utf8")
```
Comment (BlackHole1) [inline web/flat-web/scripts/vite-plugin-inline-assets.ts:19]:
```suggestion
                    ? svgToTinyDataUri(imageFileContent.toString("utf-8"))
```

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

File: web/flat-web/scripts/vite-plugin-inline-assets.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
+import type { Plugin } from "vite";
+import { promises as fsp } from "fs";
+import mime from "mime/lite";
+import svgToTinyDataUri from "@netless/mini-svg-data-uri";
+
+// e.g:
+// flat/node_modules/electron/index.js?v=19cea64f => flat/node_modules/electron/index.js
+const removeQueryString = (url: string): string => url.replace(/#.*$/, "").replace(/\?.*$/, "");
```
Comment (LitoMore) [inline web/flat-web/scripts/vite-plugin-inline-assets.ts]:
```suggestion
const removeQueryString = (url: string): string => url.replace(/[?#].*$/, "");
```

**Review by LitoMore — CHANGES_REQUESTED** [review]

File: web/flat-web/scripts/vite-plugin-inline-assets.ts (hunk 8 above)
Comment (Matrixbirds) [reply web/flat-web/scripts/vite-plugin-inline-assets.ts:19]: This function's signature is required to pass two parameters. So here need explicitly specify the arguments.

File: web/flat-web/scripts/vite-plugin-inline-assets.ts [hunk 10]
```diff
@@ -0,0 +1,28 @@
+import type { Plugin } from "vite";
+import { promises as fsp } from "fs";
+import mime from "mime/lite";
+import svgToTinyDataUri from "@netless/mini-svg-data-uri";
+
+// e.g:
+// flat/node_modules/electron/index.js?v=19cea64f => flat/node_modules/electron/index.js
+const cleanUrl = (url: string): string => url.replace(/#.*$/, "").replace(/\?.*$/, "");
```
Comment (LitoMore) [inline web/flat-web/scripts/vite-plugin-inline-assets.ts]:
```suggestion
const cleanUrl = (url: string): string => url.replace(/[?#].*$/, "");
```

**Review by LitoMore — CHANGES_REQUESTED** [review]

**Review by BlackHole1 — APPROVED** [review]
LGTM.

cc: @LitoMore please re-review

**Review by LitoMore — APPROVED** [review]

**Pull request merged**